relay-metrics/src/view.rs

```diff
@@ -0,0 +1,1061 @@
+/// A view contains a subset of datapoints of the original bucket.
+///
+/// ```txt
+///             Full Bucket
+/// /-------------------------------\
+/// [0, 1, 2, 3, 5, 5, 5, 10, 11, 11]
+/// \----------------/\-------------/
+///       View 1          View 2
+/// ```
+///
+/// A view can be split again into multiple smaller views.
+pub struct BucketView<'a> {
+    /// The source bucket.
+    inner: &'a Bucket,
+    /// Non-empty and valid range into the bucket.
+    /// The full range is constrained by `0..bucket.value.len()`
+    range: Range<usize>,
+}
+
+impl<'a> BucketView<'a> {
+    /// Creates a new bucket view of a bucket.
+    /// The resulting view contains the entire bucket.
+    pub fn new(bucket: &'a Bucket) -> Self {
+        Self {
+            inner: bucket,
+            range: 0..bucket.value.len(),
+        }
+    }
+
+    /// Timestamp of the bucket.
+    ///
+    /// See also: [`Bucket::timestamp`]
+    pub fn timestamp(&self) -> UnixTimestamp {
+        self.inner.timestamp
+    }
+
+    /// Width of the bucket.
+    ///
+    /// See also: [`Bucket::width`]
+    pub fn width(&self) -> u64 {
+        self.inner.width
+    }
+
+    /// Name of the bucket.
+    ///
+    /// See also: [`Bucket::name`]
+    pub fn name(&self) -> &str {
+        &self.inner.name
+    }
+
+    /// Value of the bucket view.
+    pub fn value(&self) -> BucketViewValue<'a> {
+        match &self.inner.value {
+            BucketValue::Counter(c) => BucketViewValue::Counter(*c),
+            BucketValue::Distribution(d) => BucketViewValue::Distribution(&d[self.range.clone()]),
+            BucketValue::Set(s) => BucketViewValue::Set(SetView::new(s, self.range.clone())),
+            BucketValue::Gauge(g) => BucketViewValue::Gauge(*g),
+        }
+    }
+
+    /// Name of the bucket.
+    ///
+    /// See also: [`Bucket::tags`]
+    pub fn tags(&self) -> &BTreeMap<String, String> {
+        &self.inner.tags
+    }
+
+    /// Returns the value of the specified tag if it exists.
+    ///
+    /// See also: [`Bucket::tag()`]
+    pub fn tag(&self, name: &str) -> Option<&str> {
+        self.inner.tag(name)
+    }
+
+    /// Number of raw datapoints in this view.
+    ///
+    /// See also: [`BucketValue::len()`]
+    pub fn len(&self) -> usize {
+        self.range.len()
+    }
+
+    /// Returns `true` if this bucket view contains no values.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Selects a sub-view of the current view.
+    ///
+    /// # Panics
+    ///
+    /// Function panics when:
+    /// * the passed range is not contained in the current view.
+    /// * trying to split a counter or gauage bucket
+    pub fn select(mut self, range: Range<usize>) -> Self {
+        assert!(
+            range.start >= self.range.start,
+            "range not contained in view"
+        );
+        assert!(range.end <= self.range.end, "range not contained in view");
+        assert!(
```

Comment:
It's a good idea to avoid panics altogether. Maybe return Option<Self> to indicate that provided range is bad or cannot be split , and issue the error.